From 6ff8065eb00d33ecc454de5f277adcad4645ca18 Mon Sep 17 00:00:00 2001
From: Ioana Alexandru <aioana@google.com>
Date: Wed, 24 Sep 2025 17:52:01 +0200
Subject: [PATCH] Enforce a hard limit for the size of images to be decoded
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Although standards like JPEG allow for sizes as large as
65,535×65,535 pixels, we realistically cannot fit such large images in
memory.

Created a test gif since it's the smallest format I could find. It's
larger than strictly necessary for the test in case we want to increase
the hard limit in the future.

Bug: 444671303
Test: manually verified that a notification that causes a crashloop
without this change, doesn't anymore
Test: LocalImageResolverTest
Flag: EXEMPT CVE_FIX

Change-Id: I0c95cfeabe169630286e7af8577c4495c2b1015f
(cherry picked from commit 449f35b532d5f680b90c8f9d8150010e7f5f30df)
Merged-In: I0c95cfeabe169630286e7af8577c4495c2b1015f
---
 .../internal/widget/LocalImageResolver.java      | 16 ++++++++++++++++
 .../internal/widget/LocalImageResolverTest.java  |  8 ++++++++
 2 files changed, 24 insertions(+)

diff --git a/core/java/com/android/internal/widget/LocalImageResolver.java b/core/java/com/android/internal/widget/LocalImageResolver.java
index 9ef7ce38fc09..a37b5eb53d2f 100644
--- a/core/java/com/android/internal/widget/LocalImageResolver.java
+++ b/core/java/com/android/internal/widget/LocalImageResolver.java
@@ -46,6 +46,12 @@ public class LocalImageResolver {
     @VisibleForTesting
     static final int DEFAULT_MAX_SAFE_ICON_SIZE_PX = 480;
 
+    /**
+     * If an image is larger than this, we won't even attempt to decode it, as we risk taking up all
+     * of the device memory.
+     */
+    private static final int DEFAULT_DECODE_HARD_LIMIT_PX = 4096;
+
     /**
      * Resolve an image from the given Uri using {@link ImageDecoder} if it contains a
      * bitmap reference.
@@ -249,6 +255,16 @@ public class LocalImageResolver {
     private static void onHeaderDecoded(ImageDecoder decoder, ImageDecoder.ImageInfo info,
             int maxWidth, int maxHeight) {
         final Size size = info.getSize();
+
+        if (size.getWidth() > DEFAULT_DECODE_HARD_LIMIT_PX
+                || size.getHeight() > DEFAULT_DECODE_HARD_LIMIT_PX) {
+            // The image is larger than what we can reasonably expect to decode without filling up
+            // the device memory, so let's bail.
+            throw new RuntimeException(
+                    "Image dimensions (" + size.getWidth() + "x" + size.getHeight()
+                            + ") exceed the maximum allowed size.");
+        }
+
         final int originalSize = Math.max(size.getHeight(), size.getWidth());
         final int maxSize = Math.max(maxWidth, maxHeight);
         final double ratio = (originalSize > maxSize)
diff --git a/core/tests/coretests/src/com/android/internal/widget/LocalImageResolverTest.java b/core/tests/coretests/src/com/android/internal/widget/LocalImageResolverTest.java
index 271a20b71106..102f8d7e0f2e 100644
--- a/core/tests/coretests/src/com/android/internal/widget/LocalImageResolverTest.java
+++ b/core/tests/coretests/src/com/android/internal/widget/LocalImageResolverTest.java
@@ -326,4 +326,12 @@ public class LocalImageResolverTest {
         Icon icon = Icon.createWithResource("invalid.package", R.drawable.test32x24);
         assertThat(LocalImageResolver.resolveResourcesForIcon(mContext, icon)).isNull();
     }
+
+    @Test(expected = IOException.class)
+    public void resolveImage_veryLargeResource_throwsException() throws IOException {
+        // Passing in an unreasonably large image should throw an exception.
+        Uri uri = Uri.parse("android.resource://"
+                + mContext.getPackageName() + "/" + R.drawable.test16000x16000);
+        LocalImageResolver.resolveImage(uri, mContext);
+    }
 }
-- 
2.34.1

